controllers/etcd_controller.go

```diff
@@ -191,86 +200,123 @@ func (r *EtcdReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 }
 
 func (r *EtcdReconciler) reconcile(ctx context.Context, etcd *druidv1alpha1.Etcd) (ctrl.Result, error) {
+	etcd, err := r.updateEtcdStatusAsNotReady(etcd, "reconciling etcd resource", druidv1alpha1.LastOperationTypeReconcile)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			return reconcileResultRequeueNone, nil
+		}
+		return reconcileResultRequeue, err
+	}
+
 	// Add Finalizers to Etcd
 	if finalizers := sets.NewString(etcd.Finalizers...); !finalizers.Has(FinalizerName) {
 		logger.Infof("Adding finalizer (%s) to etcd %s", FinalizerName, etcd.GetName())
 		finalizers.Insert(FinalizerName)
 		etcd.Finalizers = finalizers.UnsortedList()
 		if err := r.Update(ctx, etcd); err != nil {
-			if err := r.updateEtcdErrorStatus(etcd, nil, err); err != nil {
-				return ctrl.Result{
-					Requeue: true,
-				}, err
+			if err := r.updateEtcdErrorStatus(etcd, nil, err, "failed to update etcd resource", druidv1alpha1.LastOperationTypeReconcile); err != nil {
+				return reconcileResultRequeue, err
 			}
-			return ctrl.Result{
-				Requeue: true,
-			}, err
+			return reconcileResultRequeue, err
 		}
 	}
 	if err := r.addFinalizersToDependantSecrets(ctx, etcd); err != nil {
-		if err := r.updateEtcdErrorStatus(etcd, nil, err); err != nil {
-			return ctrl.Result{
-				Requeue: true,
-			}, err
+		if err := r.updateEtcdErrorStatus(etcd, nil, err, "failed to add finalizers to dependent secrets", druidv1alpha1.LastOperationTypeReconcile); err != nil {
+			return reconcileResultRequeue, err
 		}
 	}
 
-	etcd, err := r.updateEtcdStatusAsNotReady(etcd)
+	ss, err := r.getStatefulSet(ctx, etcd)
 	if err != nil {
-		if errors.IsNotFound(err) {
-			return ctrl.Result{}, nil
+		logger.Error(err, "error getting statefulset")
+		return reconcileResultRequeue, err
+	}
+
+	pvcList := &v1.PersistentVolumeClaimList{}
+	if ss != nil {
+		pvcList, err = r.listPersistentVolumeClaims(ss)
+		if err != nil {
+			return reconcileResultRequeueAfter, err
 		}
-		return ctrl.Result{
-			Requeue: true,
-		}, err
 	}
 
-	svc, ss, err := r.reconcileEtcd(etcd)
+	handleEtcdScaleDown := etcd.Spec.Replicas == 0 && len(pvcList.Items) != 0
```

Comment:
Can't this check be done inside `reconcileEtcd`? Also is the `len(pvcList.Items) != 0` check sound? Isn't a check if the backup is switched on (storage-provider?) a better check? There could be possible use-cases where we want to take snapshots even with `Local` storage provider. WDYT?